src/worm_ffi.erl

```diff
@@ -7,16 +7,11 @@ persist(Gen) ->
 	{ _, GenModule } = erlang:fun_info(Gen, module),
 	{ _, GenName } = erlang:fun_info(Gen, name),
 	Name = { ?MODULE, GenModule, GenName },
-	case catch persistent_term:get(Name) of
-		{ 'EXIT', { badarg, _ } } -> 
-			Value = Gen(),
-			try
-				persistent_term:get(Name)
-			catch
-				error:badarg:_ ->
-					persistent_term:put(Name, { ok, Value }),
-					Value
-			end;
-		{ ok, Value } -> Value
+	try
+		persistent_term:get(Name)
+	catch error:badarg ->
+		Value = Gen(),
+		persistent_term:put(Name, Value),
+		Value
```

Comment:
This is actually not as good. The reason it attempts to fetch twice is to do our best to not overwrite a PT value which would trigger a global GC pass. Since `Gen()` can take time, and erlang has real parallelism this leads to a race condition where multiple things can try to generate the cache and write to it at the same time. Refetching just before writing reduces the window for this race condition to nearly zero.

Reply:
The TL;DR from a performance perspective is that:

* `persistent_term:get(Name)` is constant time.
* `persistent_term:put(Name, Term)` for a new Name is `O(n)` where n is the number of terms stored.
* `persistent_term:put(Name, Term)` for an existing Name is `O(n)` as above, but also triggers a global GC event that will pause the whole application.

Hence, updates need to be prevented as well as we're able to. There is still technically a race condition in the current implementation, but it's an extremely small window which should be hard to hit. Your updated version increases that window to however long `Gen()` takes to run, which given the purpose of this library could be significant.

Reply:
That's a good point. I hadn't considered that. That being said my understanding is that PT doesn't actually trigger a global GC pass when the value hasn't changed. Not that that should change the approach here. I just thought it was interesting.

- https://www.erlang.org/doc/apps/erts/persistent_term.html#put/2

Reply:
Yeah, the ideal situation is that the callback should generate the same data every time it's called anyway, but it's not worth taking the risk when the extra overhead only affects the initial few calls anyways not the hot path.